### PR TITLE
Introduce base measure sampling in TimeSeriesPosterior and derived classes

### PR DESCRIPTION
**Pull request by pnkraemer**
This PR essentially
* streamlines `FiltSmoothPosterior` and `ODESolutions` under the joint interface `TimeSeriesPosterior` (formerly `FiltSmoothPosterior`) 
* introduces methods `transform_base_measure_realizations...()` in all posteriors as well as in `Transition`s. Now, the randomness in all sampling procedures lives in sample() (which is not new -- now, however, it is 1 out of 2 lines in the sample() functions, the other line delegates to `transforme_base_measure_realizations` which is where the magic happens. 
* Has default call behaviour in all posteriors. Adds the interface "interpolate()" which computes values between gridpoints.
* Updates imports in filtsmooth and diffeq
* Adds types in the modules that were touched anyway (all posterior-ones, essentially).

As a side result of this refactoring, I think I was able to simplify the logic behind dense output and sampling a bit (in terms of sorting where (a) the randomness happens, (b) the forward/backward passes happen, and all the remainders which just point to these spots).

I am aware that -- unfortunately -- the logic in the code has become quite complex. I tried my best to (i) test thoroughly, (ii) name well, and (iii) block-comment-explain lines which I thought might create confusion of sorts. Please let me know if things are not clear!

### CODE REVIEW

File: src/probnum/statespace/transition.py
Comment (pnkraemer) [inline src/probnum/statespace/transition.py:323]: In the spirit of the randomness in sampling schemes happening in `Posterior.sample()` and the "heavy lifting" being implemented here for convenience's sake, I'd propose to remove this function. The rules of how to transform the samples are captured below, and there is IMO no need for this function to be called here. Thoughts?

File: src/probnum/diffeq/odefiltsmooth/kalman_odesolution.py
Comment (pnkraemer) [inline src/probnum/diffeq/odefiltsmooth/kalman_odesolution.py:114]: wrong docs?

File: src/probnum/diffeq/odefiltsmooth/kalman_odesolution.py
Comment (pnkraemer) [inline src/probnum/diffeq/odefiltsmooth/kalman_odesolution.py:121]: explain

File: src/probnum/diffeq/odefiltsmooth/kalman_odesolution.py
Comment (pnkraemer) [inline src/probnum/diffeq/odefiltsmooth/kalman_odesolution.py:142]: kalmanposterior.sample() or kalmanposterior.transform_base_measure_realizations()?

File: src/probnum/diffeq/odesolution.py
Comment (pnkraemer) [inline src/probnum/diffeq/odesolution.py:15]: this could actually be initialised, right? So it is not really an interface?

File: src/probnum/diffeq/odesolution.py
Comment (pnkraemer) [inline src/probnum/diffeq/odesolution.py:61]: update

File: src/probnum/filtsmooth/gaussfiltsmooth/kalmanposterior.py
Comment (pnkraemer) [inline src/probnum/filtsmooth/gaussfiltsmooth/kalmanposterior.py:88]: explain?

File: tests/test_diffeq/test_odesolver.py
Comment (pnkraemer) [inline tests/test_diffeq/test_odesolver.py:32]: I think this could be replaced by `ODESolution` itself

File: src/probnum/diffeq/odefiltsmooth/kalman_odesolution.py
Comment (pnkraemer) [inline src/probnum/diffeq/odefiltsmooth/kalman_odesolution.py:132]: it should be possible to remove this argument altogether. Size should be possible to infer from the shape of base_measure_realizations?!

File: src/probnum/statespace/transition.py
Comment (pnkraemer) [reply src/probnum/statespace/transition.py]: If you agree, please let me know, and I delete this block before merging

**Comment by pnkraemer** [comment]
See also the decisions from the software meeting (#349 and #348) which are more or less resolved for these parts.

File: src/probnum/filtsmooth/timeseriesposterior.py
Comment (nathanaelbosch) [inline src/probnum/filtsmooth/timeseriesposterior.py:39]: Wasn't there a specific array-like type that we always use? I don't completely recall how it was called. But I'm also the least experienced person for the type hints, so np.ndarray might also be fine^^

File: src/probnum/filtsmooth/timeseriesposterior.py
Comment (nathanaelbosch) [inline src/probnum/filtsmooth/timeseriesposterior.py:76]: These comments should probably be removed?

File: src/probnum/filtsmooth/timeseriesposterior.py
Comment (nathanaelbosch) [inline src/probnum/filtsmooth/timeseriesposterior.py:28]: Could you briefly remind me why we need TimeSeriesPosterior, and why just having KalmanPosterior is not sufficient?

**Review by nathanaelbosch — CHANGES_REQUESTED** [review]

File: src/probnum/_randomvariablelist.py
Comment (pnkraemer) [reply src/probnum/_randomvariablelist.py:59]: Magic, right? ;)

File: src/probnum/diffeq/odefiltsmooth/kalman_odesolution.py
Comment (pnkraemer) [reply src/probnum/diffeq/odefiltsmooth/kalman_odesolution.py:123]: That is a good point. The problem is here how to slice the correct dimensions in a way that works with all kinds of samplign shapes (which I wasnt able to do, so I mirrored the KalmanPosterior logic in KalmanODESolution). This could be done with np.einsum I suppose, let me try some things and get back to rou.

File: src/probnum/filtsmooth/gaussfiltsmooth/kalmanposterior.py
Comment (pnkraemer) [reply src/probnum/filtsmooth/gaussfiltsmooth/kalmanposterior.py:68]: No, I dont think so. Seems like someone mindlessly copied the doscstring here...

File: src/probnum/filtsmooth/gaussfiltsmooth/kalmanposterior.py
Comment (pnkraemer) [reply src/probnum/filtsmooth/gaussfiltsmooth/kalmanposterior.py:249]: In principle, yes. The details are slightly different (i.e. the FilterinPosterior merely predicts, and the smoothing posterior updates as well), though. What would you prefer?

File: src/probnum/filtsmooth/gaussfiltsmooth/kalmanposterior.py
Comment (pnkraemer) [reply src/probnum/filtsmooth/gaussfiltsmooth/kalmanposterior.py:274]: I dont think this is the full point. The superclass implements `sample()` by delegationg to `transform_base_measure_...` and when trying to sample from the fitlering posterior, this is where the error would be thrown. I thought it would be a bit more human-readable if the error was thrown right at sample(), which is why it is overwritten here.
However, now that we're talking about this, it should have an error message. 

What are your thoughts?

File: src/probnum/filtsmooth/timeseriesposterior.py
Comment (pnkraemer) [reply src/probnum/filtsmooth/timeseriesposterior.py:28]: TimeSeriesPosterior is intended to be a joint interface for all sorts of posterior processes. This includes `KalmanPosterior`, but also `ODESolution` (, and `ParticleFilterPosterior`).

File: src/probnum/filtsmooth/timeseriesposterior.py
Comment (pnkraemer) [reply src/probnum/filtsmooth/timeseriesposterior.py:39]: There is numpy.typing.ArrayLike (which I think is what you're referring to), but that type is not compatible with the current numpy version we are using.

Even if we would use the ArrayLike type, I think it would only make sense a line such as `np.asarray(states)` in the first lines (which we may or may not want -- I dont know). If we specify that we want an array, this should be clear either way.

File: src/probnum/diffeq/odefiltsmooth/kalman_odesolution.py
Comment (pnkraemer) [reply src/probnum/diffeq/odefiltsmooth/kalman_odesolution.py:111]: If we delegate this entire logic here to `KalmanPosterior`, it will becovered (because this branch is tested for KalmanPosterior, but apparently not for ODE solution). If not, I will write a test for this (should be easy).

File: src/probnum/diffeq/odefiltsmooth/kalman_odesolution.py
Comment (pnkraemer) [reply src/probnum/diffeq/odefiltsmooth/kalman_odesolution.py:123]: Ignore the previous message. Here is the proper answer:

This is indeed a potentially controversial point.
We have to decide between (a) letting `KalmanODESolution` have a `transform_base_measure_realizations` which `sample()` points to, and (b) letting `KalmanODESolution.sample()` point to `KalmanPosterior.sample()` (which itself points to a `KalmanPosterior.transform_base_measure...`. 

In the spirit that `sample()` is something that "sits on top of `transform_base...`", I opted for (a). In other words, `KalmanODESolution.sample()` delegates the implementation to `KalmanPosterior` via the `transform_base_measure_realizations()` route. 

On a side note, you're right in that `KalmanODESolution.transform_base_measure_realizations` could probably do without the recursion itself. Let me see what I can do in this regard.

File: src/probnum/diffeq/odefiltsmooth/kalman_odesolution.py
Comment (pnkraemer) [reply src/probnum/diffeq/odefiltsmooth/kalman_odesolution.py:111]: Done.

File: src/probnum/diffeq/odefiltsmooth/kalman_odesolution.py
Comment (pnkraemer) [reply src/probnum/diffeq/odefiltsmooth/kalman_odesolution.py:123]: > On a side note, you're right in that KalmanODESolution.transform_base_measure_realizations could probably do without the recursion itself. Let me see what I can do in this regard.

This is taken care of.

**Comment by pnkraemer** [comment]
If you'd like to see how this functionaliy is used, check out the notebook(s) in the docs of https://github.com/probabilistic-numerics/probnum-evaluation/pull/7

File: src/probnum/filtsmooth/timeseriesposterior.py
Comment (nathanaelbosch) [reply src/probnum/filtsmooth/timeseriesposterior.py:28]: I see, thanks!

File: src/probnum/filtsmooth/gaussfiltsmooth/kalmanposterior.py
Comment (nathanaelbosch) [reply src/probnum/filtsmooth/gaussfiltsmooth/kalmanposterior.py:245]: My bad, if there are differences then of course we should have it.

File: src/probnum/diffeq/odefiltsmooth/kalman_odesolution.py
Comment (nathanaelbosch) [inline src/probnum/diffeq/odefiltsmooth/kalman_odesolution.py:108]: Could you elaborate what happens here? I know about einsums, but I don't understand why we use it here (or why it even works). Would `[_project_rv(self.proj_to_y, s) for s in samples]` do the same?

File: src/probnum/filtsmooth/gaussfiltsmooth/kalmanposterior.py
Comment (nathanaelbosch) [reply src/probnum/filtsmooth/gaussfiltsmooth/kalmanposterior.py:270]: Ah I see, makes sense. Yes I agree, an error message would be useful

File: src/probnum/statespace/transition.py
Comment (nathanaelbosch) [reply src/probnum/statespace/transition.py]: I'm not sure if this is still up to date, but in case it is I still agree :)

**Review by nathanaelbosch — APPROVED** [review]
I finally managed to go through again. Apart from a single error message to be added, I think everything should be in order! Or do you remember any other open things, from when we talked last week? I'll approve already, but if yes let me know and I can have another look.

File: src/probnum/diffeq/odefiltsmooth/kalman_odesolution.py
Comment (pnkraemer) [reply src/probnum/diffeq/odefiltsmooth/kalman_odesolution.py:108]: yes, it would do the same. Einsum is faster though, because it is vectorised :)

File: src/probnum/filtsmooth/gaussfiltsmooth/kalmanposterior.py
Comment (pnkraemer) [reply src/probnum/filtsmooth/gaussfiltsmooth/kalmanposterior.py]: Done

**Comment by pnkraemer** [comment]
Error message is added. I dont think there was anything else :) Thanks for going through again! 

On a side note: there is another tiny change in `Preconditioner`: the preconditioner is now assembled with |h| instead of h, which is something that raised weird errors for me in a different project when using ProbNum. I thought we'd get it out of the way here, I hope you dont mind :)

**Pull request merged**
